crowbar_framework/lib/crowbar/deployment_queue.rb

```diff
@@ -0,0 +1,394 @@
+
+          if proposal_queue.empty?
+            logger.debug("process queue: exit: empty queue")
+            return
+          end
+
+          logger.debug("process queue: queue: #{proposal_queue.proposals.inspect}")
+
+          # Test for ready
+          remove_list = []
+          proposal_queue.proposals.each do |item|
+            prop = Proposal.where(barclamp: item["barclamp"], name: item["inst"]).first
+
+            if prop.nil?
+              remove_list << item
+              next
+            end
+
+            next unless dependencies_satisfied?(item["deps"])
+
+            nodes_map = elements_to_nodes_to_roles_map(prop["deployment"][item["barclamp"]]["elements"],
+                                                       prop["deployment"][item["barclamp"]]["element_order"])
+            delay, pre_cached_nodes = elements_not_ready(nodes_map.keys)
+            list << item if delay.empty?
+          end
+
+          # Update the queue. Drop all proposals that we can process now (list) and those
+          # that are deleted (remove_list). This leaves in the queue only proposals
+          # which are still waiting for nodes (delay not empty), or for which deps are not
+          # ready/created/deployed (queue_me = true).
+          remove_list.each do |iii|
+            dequeue_proposal_no_lock(iii["barclamp"], iii["inst"])
+          end
+
+          list.each do |iii|
+            dequeue_proposal_no_lock(iii["barclamp"], iii["inst"])
+          end
+        rescue StandardError => e
+          logger.error("Error processing queue: #{e.message} #{e.backtrace.join("\n")}")
+          logger.debug("process queue: exit: error")
+          return
+        ensure
+          file_lock.release(f)
+        end
+
+        logger.debug("process queue: list: #{list.inspect}")
+
+        results = commit_proposals(list)
+
+        # 202 means some nodes are not ready, bail out in that case
+        # We're re-running the whole apply continuously, until there
+        # are no items left in the queue.
+        # FIXME: This is lame, because from the user perspective, we're still
+        # applying the first barclamp, while this part was in fact already
+        # completed and we're applying next item(s) in the queue.
+        loop_again = true if results.any? { |state| state != 202 }
+
+        # For each ready item, apply it.
+        logger.debug("process queue: exit")
+      end
+    end
+
+    private
+
+    def commit_proposals(list)
+      list.map do |item|
+        logger.debug("process queue: item to do: #{item.inspect}")
+
+        bc = item["barclamp"]
+        inst = item["inst"]
+
+        service = eval("#{bc.camelize}Service.new logger")
+
+        # This will call apply_role and chef-client.
+        # Params: (inst, in_queue, validate_after_save)
+        status, message = service.proposal_commit(inst, true, false)
+
+        logger.debug("process queue: item #{item.inspect}: results #{answer.inspect}")
```

Comment:
it has failed at `crowbar transition` because of that. see [here](http://pastebin.nue.suse.com/14893/src) if you are interested

other than that, I tested it and everything seems to works fine.

so after the fix above :+1: for merging